synapse/storage/roommember.py

```diff
@@ -269,3 +269,39 @@ def user_rooms_intersect(self, user_id_list):
         ret = len(room_id_lists.pop(0).intersection(*room_id_lists)) > 0
 
         defer.returnValue(ret)
+
+    def forget(self, user_id, room_id):
+        def f(txn):
+            sql = (
+                "UPDATE"
+                "  room_memberships"
+                " SET"
+                "  forgotten = 1"
+                " WHERE"
+                "  user_id = ?"
+                " AND"
+                "  room_id = ?"
+            )
+            txn.execute(sql, (user_id, room_id))
+        self.runInteraction("forget_membership", f)
+
+    @defer.inlineCallbacks
+    def did_forget(self, user_id, room_id):
+        def f(txn):
+            sql = (
+                "SELECT"
+                "  COUNT(*)"
```

Comment:
I was doing this because it's possible no rooms will be returned, and consistently handling a COUNT is a little cleaner imo than having to handle different rowcounts in python.

Now that I look at the schema, though, there are no uniqueness guarantees around room_memberships rows (which should have been obvious, I guess), so I've fixed this up to also cover the forget + re-join cases properly, and added some tests for them.